pydantic 1.8 release

## 🐛 Bug

Looks like the pydantic 1.8 release has broken something. Not sure what yet. Could make a quick pin < 1.8 or quick fix cc @tlambert03

1.8.1 just came out, not sure if it's breaking `4.6.0`, but i suspect it is breaking master due to

```
  E   TypeError: Object of type 'cycle' is not JSON serializable` error, 
```
failing in the colormanger now see https://github.com/napari/napari/pull/2344/checks?check_run_id=2026313230

cc @tlambert03 @kevinyamauchi 

not sure what could have happened from the changelog https://pydantic-docs.helpmanual.io/changelog/#v181-2021-03-03

Ok, yup I've checked this now, it is the `1.8.1` update, i havn't tried to find the PR yet, but this is the full traceback on master right now 

```python-traceback
_____________________________________________________________________________________ test_init_color_manager_colormap _____________________________________________________________________________________

    def test_init_color_manager_colormap():
        n_colors = 10
        color_cycle = [[0, 0, 0, 1], [1, 1, 1, 1]]
        properties = {'point_type': _make_cycled_properties([0, 1.5], n_colors)}
        color_manager = ColorManager._from_layer_kwargs(
            colors='point_type',
            properties=properties,
            n_colors=n_colors,
            continuous_colormap='gray',
            contrast_limits=None,
            categorical_colormap=color_cycle,
        )

        assert len(color_manager.colors) == n_colors
        assert color_manager.color_mode == 'colormap'
        color_array = transform_color(['black', 'white'] * int(n_colors / 2))
        colors = color_manager.colors.copy()
        np.testing.assert_allclose(colors, color_array)
        np.testing.assert_allclose(color_manager.current_color, [1, 1, 1, 1])
        assert color_manager.color_properties.current_value == 1.5

        # test that colormanager state can be saved and loaded
        cm_dict = color_manager.dict()
        color_manager_2 = ColorManager._from_layer_kwargs(
            colors=cm_dict, properties=properties
        )
        assert color_manager == color_manager_2

        # test json serialization
>       json_str = color_manager.json()

napari/layers/utils/_tests/test_color_manager.py:538:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pydantic/main.py:551: in pydantic.main.BaseModel.json
    ???
../../opt/anaconda3/lib/python3.7/json/__init__.py:238: in dumps
    **kw).encode(obj)
../../opt/anaconda3/lib/python3.7/json/encoder.py:199: in encode
    chunks = self.iterencode(o, _one_shot=True)
../../opt/anaconda3/lib/python3.7/json/encoder.py:257: in iterencode
    return _iterencode(o, 0)
pydantic/json.py:107: in pydantic.json.custom_pydantic_encoder
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   TypeError: Object of type 'cycle' is not JSON serializable

pydantic/json.py:95: TypeError
```

--- a/napari/utils/events/_tests/test_evented_model.py
+++ b/napari/utils/events/_tests/test_evented_model.py
@@ -284,3 +284,19 @@ def test_evented_model_serialization():
     assert raw == '{"obj": {"a": 1, "b": "hi"}, "shaped": [1.0, 2.0, 3.0]}'
     deserialized = Model.parse_raw(raw)
     assert deserialized == m
+
+
+def test_nested_evented_model_serialization():
+    """Test that encoders on nested sub-models can be used by top model."""
+
+    class NestedModel(EventedModel):
+        obj: MyObj
+
+    class Model(EventedModel):
+        nest: NestedModel
+
+    m = Model(nest={'obj': {"a": 1, "b": "hi"}})
+    raw = m.json()
+    assert raw == r'{"nest": {"obj": {"a": 1, "b": "hi"}}}'
+    deserialized = Model.parse_raw(raw)
+    assert deserialized == m

--- a/napari/utils/events/evented_model.py
+++ b/napari/utils/events/evented_model.py
@@ -84,7 +84,12 @@ class EventedMetaclass(main.ModelMetaclass):
             # NOTE: a _json_encode field must return an object that can be
             # passed to json.dumps ... but it needn't return a string.
             if hasattr(f.type_, '_json_encode'):
-                cls.__config__.json_encoders[f.type_] = f.type_._json_encode
+                encoder = f.type_._json_encode
+                cls.__config__.json_encoders[f.type_] = encoder
+                # also add it to the base config
+                # required for pydantic>=1.8.0 due to:
+                # https://github.com/samuelcolvin/pydantic/pull/2064
+                EventedModel.__config__.json_encoders[f.type_] = encoder
         return cls
 
 

